File: frontend/src/types/trace.ts
// Mirror of the backend trace vocabulary (app/core/events.py). Order + timing
// are the signal, so every event carries a monotonic seq and a client timestamp.

export type EventType =
  | "session_started"
  | "assessment_started"
  | "agent_note_viewed"
  | "file_opened"
  | "file_edited"
  | "code_run"
  | "test_added"
  | "test_edited"
  | "test_run"
  | "marked_ready"
  | "session_ended"
  | "timed_out";

// At ship time we ask whether the candidate verified the agent's "I tested it"
// claim. The answer is checked against the trace (e.g. "ran_tests" with no
// test_run event is a tell).
export type VerifyClaim = "ran_tests" | "read_not_run" | "trusted";

export const VERIFY_CLAIM_LABELS: Record<VerifyClaim, string> = {
  ran_tests: "Ran the tests myself",
  read_not_run: "Read the tests but didn't run them",
  trusted: "Trusted the agent's note",
};

// The submit decision: is this change ready to ship, and how sure are they?
export type Decision = "ship" | "ship_with_caveats" | "block";
export type Confidence = "low" | "medium" | "high";

export const DECISION_LABELS: Record<Decision, string> = {
  ship: "Ship it",
  ship_with_caveats: "Ship with caveats",
  block: "Block — not ready",
};

export const CONFIDENCE_LABELS: Record<Confidence, string> = {
  low: "Low",
  medium: "Medium",
  high: "High",
};

// One executed test case, captured from the Tests panel result tree.
export interface TestResult {
  name: string;
  status: "pass" | "fail";
}

// Per-type payload shapes (loose by design — the scorer reads these by hand).
export interface EventPayloads {
  session_started: { task_id: string; candidate_name: string };
  assessment_started: Record<string, never>;
  agent_note_viewed: { dwell_ms: number };
  file_opened: { path: string };
  // content is the full-file snapshot; the backend stamps a `diff` onto the
  // stored payload at ingest (the client doesn't send it).
  file_edited: { path: string; content: string };
  // The test file is edited via its own event, kept distinct from file_edited.
  test_edited: { path: string; content: string };
  // Verbatim scratchpad source + console output. Output is capped (truncated
  // flag set) so a runaway loop can't bloat the trace.
  code_run: { command: string; source: string; output: string; truncated: boolean };
  test_added: { path: string };
  test_run: {
    result: { passed: number; failed: number; total: number };
    tests: TestResult[]; // which tests ran, and their outcome
  };
  marked_ready: {
    files: Record<string, string>;
    // Structured trust-vs-verify answer, cross-checkable against the trace.
    verify_claim: VerifyClaim;
    // The ship/caveat/block decision + how confident they are.
    decision: Decision;
    confidence: Confidence;
    // Required neutral hand-off note — the primary evidence for the Scope moments.
    team_summary: string;
    // Free-text: what they changed / checked. Optional.
    reflection: string;
  };
  session_ended: Record<string, never>;
  timed_out: Record<string, never>;
}

export interface TraceEventInput<T extends EventType = EventType> {
  seq: number;
  client_ts: string; // ISO 8601
  type: T;
  payload: EventPayloads[T];
}

// ---- Dashboard read models (mirror backend app/models/trace.py) ----

export interface SessionSummary {
  id: string;
  candidate_name: string;
  task_id: string;
  status: string;
  marked_ready: boolean;
  created_at: string;
  started_at: string | null;
  ended_at: string | null;
}

export interface TraceEventOut {
  id: string;
  created_at: string;
  seq: number;
  client_ts: string;
  type: EventType;
  payload: Record<string, unknown>;
}

export interface FileDiff {
  path: string;
  unified: string;
}

export interface Session extends SessionSummary {
  updated_at: string;
  final_files: Record<string, string> | null;
  decision: Decision | null;
  confidence: Confidence | null;
  team_summary: string | null;
}

export interface TraceExport {
  session: Session;
  events: TraceEventOut[];
  file_diffs: FileDiff[];
}

// ---- Analysis (mirror backend SessionAnalysis) ----

export interface VerificationCheck {
  key: string;
  label: string;
  status: "yes" | "no" | "warn";
  detail: string;
  evidence_seq: number | null;
}

export interface MomentSignal {
  key: string;
  title: string;
  signal: "missed" | "noticed" | "acted";
  evidence: string;
}

export interface CodeComparison {
  path: string;
  original: string;
  final: string;
  unified: string;
}

export interface SelfReport {
  verify_claim: string | null;
  verify_claim_label: string;
  reflection: string;
  decision: Decision | null;
  decision_label: string;
  confidence: Confidence | null;
  team_summary: string;
}

export interface SessionAnalysis {
  session: Session;
  verification: VerificationCheck[];
  moments: MomentSignal[];
  code: CodeComparison[];
  self_report: SelfReport;
}

export interface ComparisonRow {
  id: string;
  candidate_name: string;
  status: string;
  created_at: string;
  verify_claim_label: string;
  reflection: string;
  decision: Decision | null;
  decision_label: string;
  confidence: Confidence | null;
  team_summary: string;
  verification: VerificationCheck[];
  moments: MomentSignal[];
}

File: backend/app/models/trace.py
"""Request/response shapes for tasks, sessions, and trace events."""

from datetime import datetime

from pydantic import BaseModel, Field

from app.core.events import EventType


# ---- Tasks -----------------------------------------------------------------

class TaskManifest(BaseModel):
    """What the candidate sees: the editable files plus the framing text.

    Deliberately omits the answer key — see task_loader, which serves only
    files on the manifest allowlist.
    """

    task_id: str
    prompt: str                      # TASK.md
    agent_note: str                  # AGENT_NOTE.md
    files: dict[str, str]            # {sandbox_path: content}, e.g. "src/cart.ts"
    entry: str                       # path of the file to open first


# ---- Sessions --------------------------------------------------------------

class SessionCreate(BaseModel):
    task_id: str
    candidate_name: str = Field(min_length=1)


class SessionCreated(BaseModel):
    session_id: str


class Session(BaseModel):
    id: str
    created_at: datetime
    updated_at: datetime
    task_id: str
    candidate_name: str
    started_at: datetime | None = None
    ended_at: datetime | None = None
    status: str
    marked_ready: bool
    final_files: dict[str, str] | None = None
    # Structured submit decision (null until the candidate submits).
    decision: str | None = None        # "ship" | "ship_with_caveats" | "block"
    confidence: str | None = None      # "low" | "medium" | "high"
    team_summary: str | None = None    # required free-text at submit


# ---- Trace events ----------------------------------------------------------

class TraceEventIn(BaseModel):
    """One buffered event from the browser. seq is monotonic per session."""

    seq: int = Field(ge=0)
    client_ts: datetime
    type: EventType
    payload: dict = Field(default_factory=dict)


class EventBatch(BaseModel):
    events: list[TraceEventIn] = Field(min_length=1)


class TraceEventOut(TraceEventIn):
    id: str
    created_at: datetime


class FileDiff(BaseModel):
    path: str
    unified: str                     # unified diff vs the previous snapshot


class TraceExport(BaseModel):
    """Everything needed to hand-score one session."""

    session: Session
    events: list[TraceEventOut]
    file_diffs: list[FileDiff]       # computed from consecutive file_edited snapshots


class VerificationCheck(BaseModel):
    key: str
    label: str
    status: str            # "yes" | "no" | "warn"
    detail: str = ""
    evidence_seq: int | None = None  # the trace event that justifies it


class MomentSignal(BaseModel):
    key: str
    title: str
    signal: str            # "missed" | "noticed" | "acted" — a HINT, not a score
    evidence: str = ""


class CodeComparison(BaseModel):
    path: str
    original: str          # the agent's code
    final: str             # what the candidate shipped
    unified: str           # unified diff agent -> shipped


class SelfReport(BaseModel):
    verify_claim: str | None = None
    verify_claim_label: str = ""
    reflection: str = ""
    # The structured submit decision + its primary free-text evidence.
    decision: str | None = None
    decision_label: str = ""
    confidence: str | None = None
    team_summary: str = ""


class SessionAnalysis(BaseModel):
    """Computed scoring aids for the dashboard. Heuristic hints; the human
    assigns the final ladder score (v1 scoring is manual)."""

    session: Session
    verification: list[VerificationCheck]
    moments: list[MomentSignal]
    code: list[CodeComparison]
    self_report: SelfReport


class ComparisonRow(BaseModel):
    """One candidate's compact scoring profile, for the side-by-side compare view."""

    id: str
    candidate_name: str
    status: str
    created_at: datetime
    verify_claim_label: str
    reflection: str
    decision: str | None = None
    decision_label: str = ""
    confidence: str | None = None
    team_summary: str = ""
    verification: list[VerificationCheck]
    moments: list[MomentSignal]


class SessionSummary(BaseModel):
    id: str
    candidate_name: str
    task_id: str
    status: str
    marked_ready: bool
    created_at: datetime
    started_at: datetime | None = None
    ended_at: datetime | None = None
    decision: str | None = None
